cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/MaintenanceEventDetector.java

```diff
@@ -6,42 +6,112 @@
 
 import com.linkedin.cruisecontrol.detector.Anomaly;
 import com.linkedin.kafka.cruisecontrol.KafkaCruiseControl;
+import com.linkedin.kafka.cruisecontrol.common.KafkaCruiseControlThreadFactory;
 import com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig;
+import com.linkedin.kafka.cruisecontrol.config.constants.AnomalyDetectorConfig;
 import java.time.Duration;
 import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Queue;
 import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import static com.linkedin.kafka.cruisecontrol.config.constants.AnomalyDetectorConfig.MAINTENANCE_EVENT_READER_CLASS_CONFIG;
 import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.getAnomalyDetectionStatus;
 import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.KAFKA_CRUISE_CONTROL_OBJECT_CONFIG;
 
 
 public class MaintenanceEventDetector extends AbstractAnomalyDetector implements Runnable {
   private static final Logger LOG = LoggerFactory.getLogger(MaintenanceEventDetector.class);
   public static final long DETECTION_NOT_READY_BACKOFF_MS = TimeUnit.SECONDS.toMillis(10);
+  public static final long IDEMPOTENCE_CACHE_CLEANER_PERIOD_SECONDS = 5;
+  public static final long IDEMPOTENCE_CACHE_CLEANER_INITIAL_DELAY_SECONDS = 0;
   // TODO: Make this configurable.
   public static final Duration READ_EVENTS_TIMEOUT = Duration.ofSeconds(5);
   private volatile boolean _shutdown;
   private final MaintenanceEventReader _maintenanceEventReader;
+  private final boolean _idempotenceEnabled;
+  private final Duration _idempotenceRetention;
+  private final ScheduledExecutorService _idempotenceCacheCleaner;
+  private final Map<Long, MaintenanceEvent> _maintenanceEventByTime;
```

Comment:
Instead of storing reference of `MaintenanceEvent` as values in the map, can we just add `hashCode()` and `equals()` methods to the `MaintenanceEvent` class and store hash code of `MaintenanceEvent` object in this map? I think we can save memory and increase the max cache size by a lot

Reply:
(_This patch already adds `hashCode()` and `equals()` methods to the `MaintenanceEvent` class_).
I see the reasoning, but as discussed earlier this cache is not expected to contain many entries. Hence, I don't think the additional programming complexity of storing hashcode rather than the event itself would worth it.